// @react-md/utils/lib/wia-aria/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var FocusContainer_1 = require("./FocusContainer");
Object.defineProperty(exports, "FocusContainer", { enumerable: true, get: function () { return FocusContainer_1.default; } });
__exportStar(require("./FocusContainer"), exports);
__exportStar(require("./movement"), exports);
var useScrollLock_1 = require("./useScrollLock");
Object.defineProperty(exports, "useScrollLock", { enumerable: true, get: function () { return useScrollLock_1.default; } });
Object.defineProperty(exports, "DATA_RMD_NOSCROLL", { enumerable: true, get: function () { return useScrollLock_1.DATA_RMD_NOSCROLL; } });
var useFocusOnMount_1 = require("./useFocusOnMount");
Object.defineProperty(exports, "useFocusOnMount", { enumerable: true, get: function () { return useFocusOnMount_1.default; } });
var usePreviousFocus_1 = require("./usePreviousFocus");
Object.defineProperty(exports, "usePreviousFocus", { enumerable: true, get: function () { return usePreviousFocus_1.default; } });
var useCloseOnEscape_1 = require("./useCloseOnEscape");
Object.defineProperty(exports, "useCloseOnEscape", { enumerable: true, get: function () { return useCloseOnEscape_1.default; } });
var getFocusableElements_1 = require("./getFocusableElements");
Object.defineProperty(exports, "getFocusableElements", { enumerable: true, get: function () { return getFocusableElements_1.default; } });
var focusElementWithin_1 = require("./focusElementWithin");
Object.defineProperty(exports, "focusElementWithin", { enumerable: true, get: function () { return focusElementWithin_1.default; } });
var extractTextContent_1 = require("./extractTextContent");
Object.defineProperty(exports, "extractTextContent", { enumerable: true, get: function () { return extractTextContent_1.default; } });
